rosidl_typesupport_introspection_tests/test/introspection_libraries_under_test.hpp

```diff
@@ -0,0 +1,620 @@
+
+#include <memory>
+
+#include <rosidl_typesupport_introspection_cpp/message_introspection.hpp>
+#include <rosidl_typesupport_introspection_cpp/service_introspection.hpp>
+
+#include "rosidl_typesupport_introspection_tests/msg/arrays.h"
+#include "rosidl_typesupport_introspection_tests/msg/basic_types.h"
+#include "rosidl_typesupport_introspection_tests/msg/bounded_sequences.h"
+#include "rosidl_typesupport_introspection_tests/msg/constants.h"
+#include "rosidl_typesupport_introspection_tests/msg/defaults.h"
+#include "rosidl_typesupport_introspection_tests/msg/empty.h"
+#include "rosidl_typesupport_introspection_tests/msg/multi_nested.h"
+#include "rosidl_typesupport_introspection_tests/msg/strings.h"
+#include "rosidl_typesupport_introspection_tests/msg/unbounded_sequences.h"
+
+#include "rosidl_typesupport_introspection_tests/msg/arrays.hpp"
+#include "rosidl_typesupport_introspection_tests/msg/basic_types.hpp"
+#include "rosidl_typesupport_introspection_tests/msg/bounded_sequences.hpp"
+#include "rosidl_typesupport_introspection_tests/msg/constants.hpp"
+#include "rosidl_typesupport_introspection_tests/msg/defaults.hpp"
+#include "rosidl_typesupport_introspection_tests/msg/empty.hpp"
+#include "rosidl_typesupport_introspection_tests/msg/multi_nested.hpp"
+#include "rosidl_typesupport_introspection_tests/msg/strings.hpp"
+#include "rosidl_typesupport_introspection_tests/msg/unbounded_sequences.hpp"
+
+#include "rosidl_typesupport_introspection_tests/fixtures.hpp"
+#include "rosidl_typesupport_introspection_tests/helpers.hpp"
+#include "rosidl_typesupport_introspection_tests/libraries.hpp"
+#include "rosidl_typesupport_introspection_tests/type_traits.hpp"
+
+namespace rosidl_typesupport_introspection_tests
+{
+
+// Typesupport library definition for introspection of test interfaces in C
+struct IntrospectionCTypeSupportTestLibrary
+{
+  using MessageDescriptorT =
+    rosidl_typesupport_introspection_c__MessageMembers;
+  using ServiceDescriptorT =
+    rosidl_typesupport_introspection_c__ServiceMembers;
+  using MemberDescriptorT =
+    rosidl_typesupport_introspection_c__MessageMember;
+
+  static constexpr const char * name = RCUTILS_STRINGIFY(
+    ROSIDL_TYPESUPPORT_INTERFACE__LIBRARY_NAME(
+      rosidl_typesupport_introspection_c,
+      rosidl_typesupport_introspection_tests));
+  static constexpr const char * identifier =
+    "rosidl_typesupport_introspection_c";
+
+  static constexpr const char * messages_namespace =
+    "rosidl_typesupport_introspection_tests__msg";
+  static constexpr const char * services_namespace =
+    "rosidl_typesupport_introspection_tests__srv";
+  static constexpr const char * actions_namespace =
+    "rosidl_typesupport_introspection_tests__action";
+
+  static constexpr const MessageTypeSupportSymbolRecord messages[] = {
+    MESSAGE_TYPESUPPORT_SYMBOL_RECORD(
+      rosidl_typesupport_introspection_c,
+      rosidl_typesupport_introspection_tests, msg, Arrays),
```

Comment:
Not sure I follow you here.

Reply:
Sorry, let me explain it:

In https://github.com/ros2/rosidl/pull/651/files#diff-31e6f9e3f2b7b5c3a3250552def63bc483cbbcd37ccd184b226445523af638d1R82-R108 we are calling 
```c++
 MESSAGE_TYPESUPPORT_SYMBOL_RECORD(
      rosidl_typesupport_introspection_c,
      rosidl_typesupport_introspection_tests, msg, $$VALUE$$),
``` 
Where `$$VALUE$$` is in the list of "Array, BasicTypes, BoundedSequences, ...". Not sure if it would be possible to replace the duplication with a loop somehow.

Just below https://github.com/ros2/rosidl/pull/651/files#diff-31e6f9e3f2b7b5c3a3250552def63bc483cbbcd37ccd184b226445523af638d1R122-R205, for the traits we are doing the same by duplicating:

```c++
template<>
struct introspection_traits<rosidl_typesupport_introspection_tests__msg__Arrays>
{
  static constexpr const MessageTypeSupportSymbolRecord typesupport =
    MESSAGE_TYPESUPPORT_SYMBOL_RECORD(
    rosidl_typesupport_introspection_c,
    rosidl_typesupport_introspection_tests, msg, $$VALUE$$);
  using TypeSupportLibraryT = IntrospectionCTypeSupportTestLibrary;
};
```

Is not that I have something clear in mind to reduce the duplication, just asking. Feel free to drop if you no easy way.

Reply:
Oh I see. 

> Not sure if it would be possible to replace the duplication with a loop somehow.

These are compile time definitions, so looping isn't possible. We could add some variadic macros to recurse though. Duplication will go away but the macros are tricky.

> for the traits we are doing the same by duplicating:

Right. Same answer. It is doable with variadic macros.

Reply:
Thanks for the feedback, proposal was not a blocker for this PR.